featurebyte/utils/async_helper.py

```diff
@@ -31,11 +150,13 @@ def asyncio_gather(*coros: Coroutine[Any, Any, Any], max_concurrency: int = 0) -
     failed = False
     tasks_canceled = False
     tasks = []
-    semaphore = asyncio.Semaphore(max_concurrency)
+    semaphore_key = f"asyncio_gather:{concurrency_key}"
 
     # Run coroutines with a semaphore
     async def _coro(coro: Coroutine[Any, Any, Any]) -> Any:
-        async with semaphore:
+        async with RedisFairCountingSemaphore(
+            redis=redis, key=semaphore_key, limit=max_concurrency, timeout=3600
```

Comment:
Does this mean that a query will be queued for maximum of 1 hour by default before it raises an error? Is this a new behaviour compared to the existing behaviour?

Reply:
In both the current and new behaviour there is a limit on how many SQL commands can be executed at the same time. 

Current behaviour: 
- the limit is local and more commands can be executed in parallel via multiple worker tasks
- no timeout on semaphore (since it's local) but there should be timeout from the calling function or the worker task

New behaviour: t
- the limit is cluster-wide and specific to each data warehouse
- timeout on semaphore to avoid situations where a process with the semaphore is terminated without releasing the semaphore indefinitely
- timeout from the calling function or the worker task still applies and will trigger release of the semaphore acquired

We can set the semaphore timeout to a value that is larger than expected higher level timeout limits to keep current behaviour